object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/ObjectConstructionAnnotatedTypeFactory.java

```diff
@@ -554,6 +576,66 @@ private void mustCallTraverse(ControlFlowGraph cfg) {
     }
   }
 
+  private void checkAssignmentToOwningField(AssignmentNode assignmentNode) {
+    Node lhs = assignmentNode.getTarget();
+    Node rhs = assignmentNode.getExpression();
+    Element fieldElement = TreeUtils.elementFromTree(lhs.getTree());
+    Element enclosingElement = fieldElement.getEnclosingElement();
+    List<String> rhsMCAnno = getMustCallValue(TreeUtils.elementFromTree(rhs.getTree()));
+    AnnotationMirror rhsCalledMethodAnno = createCalledMethods(rhsMCAnno.toArray(new String[0]));
+    List<String> enclosingElAnno = getMustCallValue(enclosingElement);
+    boolean report = true;
+    if (enclosingElAnno != null
+        && (getMustCallValue(fieldElement).isEmpty()
+            || fieldElement.getModifiers().contains(Modifier.FINAL))) {
+      List<? extends Element> classElements = enclosingElement.getEnclosedElements();
+      for (Element element : classElements) {
+        if (element.getKind().equals(ElementKind.METHOD)) {
+          // TODO cause false positives if there is more than one value in the enclosingElAnno
+          if (element.getSimpleName().toString().equals(enclosingElAnno.get(0))) {
+            AnnotationMirror annotationMirror =
+                getDeclAnnotation(element, EnsuresCalledMethods.class);
+            if (annotationMirror == null) {
+              break;
+            }
+            Attribute values = ((Attribute.Compound) annotationMirror).getValue().values.get(0).snd;
+            if (values.getValue().toString().contains(fieldElement.getSimpleName().toString())) {
+              Attribute methods =
+                  ((Attribute.Compound) annotationMirror).getValue().values.get(1).snd;
+              Set<String> setOfMethod =
+                  ((List<Attribute>) methods.getValue())
+                      .stream()
+                          .map(method -> method.getValue().toString())
+                          .collect(Collectors.toSet());
+              AnnotationMirror ensuresCalledMethodAnno =
+                  createCalledMethods(setOfMethod.toArray(new String[0]));
+              if (getQualifierHierarchy().isSubtype(ensuresCalledMethodAnno, rhsCalledMethodAnno)) {
```

Comment:
Why is this doing a check comparing against `rhsCalledMethodAnno`?  I think we should only be checking that the `@EnsuresCalledMethod` annotation has all methods in the `@MustCall` annotation for the _declared type_ of the field.  It should be independent of any assignment into the field.